server/oauth2.go

```diff
@@ -572,12 +572,15 @@ func (s *Server) validateCrossClientTrust(clientID, peerID string) (trusted bool
 }
 
 func validateRedirectURI(client storage.Client, redirectURI string) bool {
-	if !client.Public {
-		for _, uri := range client.RedirectURIs {
-			if redirectURI == uri {
-				return true
-			}
+	// Allow named RedirectURIs for both public and non-public clients.
+	// This is required make PKCE-enabled web apps work, when configured as public clients.
+	for _, uri := range client.RedirectURIs {
+		if redirectURI == uri {
+			return true
 		}
+	}
+	// For non-public clients, only named RedirectURIs are allowed.
+	if !client.Public {
 		return false
 	}
```

Comment:
Our use case is to use PKCE for a web app that doesn't run on "localhost".
Therefore, we would like that `RedirectURIs` are also considered for public clients.
PKCE was definitely also intended for web apps, and you can already configure such web apps e.g. in Okta or Auth0.

The following code to allow any localhost URI, `redirectURIOOB` or `deviceCallbackURI` for public clients is still active.
An alternative could be to no longer allow these special URLs implicitly when `RedirectURIs` is non-empty.
I don't have a strong opinion if this would be better. The maintainers should decide.

Reply:
I am curious - why don't you configure the app as confidential client if it is a web app and can keep the client secret secure?

Reply:
I think the implementation is good as it is. The alternative of not supporting the special URLs, when `redirectURIs` is defined, would break currently working configurations.

Reply:
I would personally prefer the logic to only allow whitelisted redirect uris, if the list is non-empty.

This would tighten the security and would not break existing configurations that do not use RedirectURIs for public clients.
And I bet the most configurations that do currently use it for public clients operate under the false assumption that the list is actually respected (I know we did in our deployment for some time).

Reply:
I fear that there are configurations that may be working by accident, because currently it is possible to have the RedirectURIs wrongly configured, and it still works.

Let's see what the maintainers decide.

Reply:
> I would personally prefer the logic to only allow whitelisted redirect uris, if the list is non-empty.

@tkleczek AFAICT this is not the case now, but the rest of the conditions provide backwards compatibility and limit redirect uris to `redirectURIOOB`, `deviceCallbackURI` or `localhost` for public clients which seems fine. WDYT?

Reply:
My concern is that for SPAs allowing localhost redirects seems like an unnecessary risk.
But if we care about potential backwards compatibility issues more, then I am fine with the change as it is.

Reply:
> My concern is that for SPAs allowing localhost redirects seems like an unnecessary risk.

That's actually true. Consider that e.g. a malicious desktop app could obtain a token intended for https://webapp.com , since the desktop app can run a server listening on http://localhost .
If the user is already logged in at the IDP in the browser, then this could probably be done without user interaction.
The desktop app could then call the API of https://webapp.com with the token.

So I would prefer not to allow these special URLs implicitly if `RedirectURIs` is non-empty.

In the future, three additional boolean options could be added to explicitly (dis)allow each of `redirectURIOOB`, `deviceCallbackURI` or `localhost`-ish URLs. But this would definitely break backwards compatibility, and might be over-engineering.

Reply:
@sagikazarmark I added more tests for these special URIs in https://github.com/dexidp/dex/pull/1822/commits/c15e2887bc262a7dae2df741c7ba06f866272272 .  
Then I made the change suggested by @tkleczek in https://github.com/dexidp/dex/pull/1822/commits/162073b33e46e502696b47ee97f871b944540114 : `if !client.Public || len(client.RedirectURIs) > 0 { return false }`

Please take another look.